mailbox/store/src/main/java/org/apache/james/mailbox/store/search/SearchGuessingAlgorithmImpl.java

```diff
@@ -0,0 +1,108 @@
+ * software distributed under the License is distributed on an    *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY         *
+ * KIND, either express or implied.  See the License for the      *
+ * specific language governing permissions and limitations        *
+ * under the License.                                             *
+ ******************************************************************/
+
+package org.apache.james.mailbox.store.search;
+
+import java.util.Collections;
+
+import javax.inject.Inject;
+
+import org.apache.james.mailbox.MailboxSession;
+import org.apache.james.mailbox.exception.MailboxException;
+import org.apache.james.mailbox.model.Mailbox;
+import org.apache.james.mailbox.model.MailboxACL;
+import org.apache.james.mailbox.model.MailboxId;
+import org.apache.james.mailbox.model.MessageId;
+import org.apache.james.mailbox.model.MultimailboxesSearchQuery;
+import org.apache.james.mailbox.store.MailboxSessionMapperFactory;
+import org.apache.james.mailbox.store.StoreRightManager;
+import org.apache.james.mailbox.store.mail.MailboxMapper;
+import org.apache.james.mailbox.store.mail.MessageIdMapper;
+import org.apache.james.mailbox.store.mail.MessageMapper;
+import org.apache.james.mailbox.store.mail.model.MailboxMessage;
+
+import com.github.fge.lambdas.Throwing;
+import com.github.steveash.guavate.Guavate;
+import com.google.common.collect.ImmutableSet;
+
+import reactor.core.publisher.Flux;
+
+public class SearchGuessingAlgorithmImpl implements SearchGuessingAlgorithm {
+    private final MessageSearchIndex index;
+    private final MailboxSessionMapperFactory mailboxSessionMapperFactory;
+    private final StoreRightManager storeRightManager;
+
+    @Inject
+    public SearchGuessingAlgorithmImpl(MessageSearchIndex index, MailboxSessionMapperFactory mailboxSessionMapperFactory, StoreRightManager storeRightManager) {
+        this.index = index;
+        this.mailboxSessionMapperFactory = mailboxSessionMapperFactory;
+        this.storeRightManager = storeRightManager;
+    }
+
+    @Override
+    public Flux<MailboxMessage> searchMailboxMessages(MultimailboxesSearchQuery expression, MailboxSession session, long limit, MessageMapper.FetchType fetchType) throws MailboxException {
+        MessageIdMapper messageIdMapper = mailboxSessionMapperFactory.getMessageIdMapper(session);
+
+        return searchMessageIds(expression, session, limit)
+            .flatMap(messageId -> messageIdMapper.findReactive(Collections.singletonList(messageId), fetchType));
+    }
+
+    @Override
+    public Flux<MessageId> searchMessageIds(MultimailboxesSearchQuery expression, MailboxSession session, long limit) throws MailboxException {
+        return getInMailboxIds(expression, session)
+            .filter(id -> !expression.getNotInMailboxes().contains(id))
+            .collect(Guavate.toImmutableSet())
+            .flatMapMany(Throwing.function(ids -> index.search(session, ids, expression.getSearchQuery(), limit)));
```

Comment:
Use mailboxManager for the search...

Reply:
That would be a loop: MailboxManager need ThreadIdGuessingAlgorithm and ThreadIdGuessingAlgorithm need MailboxManager

Reply:
Injection systems like guice are designed to break these loops.